driver-core/src/main/com/mongodb/operation/QueryBatchCursor.java

```diff
@@ -106,6 +107,17 @@
             this.connectionSource = null;
         }
     }
+
+    private boolean canConnectionBeExhaust() {
+        Connection connection = connectionSource.getConnection();
+        if (serverIsAtLeastVersionFourDotOne(connection.getDescription())) {
```

Comment:
If this is false, it will leak the connection from the line above.  I'm surprised the tests running on version <= 4.0 didn't catch this.

What you can do instead is just check the connection that's passed in to the constructor instead of getting a new one from the connection source.  If it's >= 4.1, then

    exhaustConnection = connection.retain();

Reply:
okay, I was thinking of doing that. I just wasn't sure if that connection would ever be null?

Reply:
It won't be.